lib/HLSL/DxilValidation.cpp

```diff
@@ -1131,6 +1131,54 @@ static int GetCBufSize(Value *cbHandle, ValidationContext &ValCtx) {
   return RP.CBufferSizeInBytes;
 }
 
+// Make sure none of the handle arguments are undef / zero-initializer,
+// Also, do not accept any resource handles with invalid dxil resource
+// properties
+void ValidateHandleArgsForInstruction(CallInst *CI, DXIL::OpCode opcode,
+                                      ValidationContext &ValCtx) {
+
+  for (Value *op : CI->operands()) {
+    const Type *pHandleTy = ValCtx.HandleTy; // This is a resource handle
+    const Type *pNodeHandleTy = ValCtx.DxilMod.GetOP()->GetNodeHandleType();
+    const Type *pNodeRecordHandleTy =
+        ValCtx.DxilMod.GetOP()->GetNodeRecordHandleType();
+
+    const Type *argTy = op->getType();
+    if (argTy == pNodeHandleTy || argTy == pNodeRecordHandleTy ||
+        argTy == pHandleTy) {
+      bool bIsUninitialized =
+          isa<UndefValue>(op) || isa<ConstantAggregateZero>(op);
+      if (argTy == pHandleTy) {
+        hlsl::DxilResourceProperties DRP = GetResourceFromHandle(op, ValCtx);
+        if (!DRP.isValid()) {
+          ValCtx.EmitInstrError(CI, ValidationRule::InstrResourceInvalid);
```

Comment:
My point from the previous comment here was that using `GetResourceFromHandle` will emit an error on an invalid handle, including the case you're checking here, without the need for an additional check.  You commented that `GetResourceFromHandle` checks ResourceClass, but the `getResourceClass` method it uses actually checks the `ResourceKind` and will return `ResourceClass::Invalid` when `ResourceKind::Invalid`, so there is no gap.  This means that the new error will always be emitted after another error has already been emitted from `GetResourceFromHandle`.  I don't think adding the new error is necessary.  A comment above the call to `GetResourceFromHandle` could indicate that the code expects that call to emit an error when the handle is not valid (like: "GetResourceFromHandle will emit an error on an invalid handle").

I also think that for the zeroinit/undef case, we could universally emit the error, before checking which handle type we have.  That way, this error will be consistent for all zeroinit or undef handle usage.  You can just change the initialization of `bIsUninitialized` to an if that emits this error, with an `else if` for the additional resource handle checks.

This may require test updates, if there are any tests that exercise the zeroinit or undef handle case for resources.

Another issue I just thought of:
For invalid resource handles, if/when there is validation of the operation itself, usually this will require retrieval of the handle, often using `GetResourceFromHandle` again.  This means any errors emitted there will now be emitted twice for these operations.

So, while this change covers potential gaps in invalid handle usage in DXIL ops, it also duplicates the coverage where this was already detected later.  The current way that some things flow through to `GetResourceFromHandle` and emit errors, while some others use `GetResourceFromVal` which emits no errors, is currently a mess.

To resolve this, I think the basic handle validation can be up-front, like this, to guarantee that it always emits a consistent error for invalid handles.  Then, when looking up handle properties in the individual operation validation cases, those should be using `GetResourceFromVal()` and `IsValid()` to skip the ResourceProperties-based checks on invalid handles, to avoid spurious additional errors when an error will have already been emitted for an invalid handle.

I also don't think the distinction between `GetResourceFromHandle` and `GetResourceFromVal` name-wise and functionality-wise makes sense.  The first should be changed into a validation-only check with a new name, such as `ValidateResourceHandle()`.  Then this would be called under this basic handle validation pass only.  Other uses of `GetResourceFromHandle` would be changed to use `GetResourceFromVal` instead.  Then all uses of `GetResourceFromVal` should be checked to make sure any uses of those resource properties for additional validation are placed under control flow guarded by `IsValid()` to avoid the additional spurious errors.

Note that `GetResourceFromHandle` is used in functions for getting specific properties (`GetSamplerKind`, `GetResourceKindAndCompTy`, `GetCBufSize`).  Those functions should probably DXASSERT if the handle is invalid, since callers should guard against that before using them.  Then the callers need new code to call `GetResourceFromVal` and `IsValid()` to guard those code paths.  Note that `ValidateSampleInst` and `ValidateGather` could share the code that calls `GetSamplerKind` (and checks for valid handle) with a tiny refactor.

The additional changes I suggest could be done as part of this change, or we could file an additional issue to track that work and follow up with a later change instead.